pkg/controllers/cinder/cinder.go

```diff
@@ -0,0 +1,151 @@
+package cinder
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gophercloud/gophercloud/openstack/blockstorage/v3/volumetypes"
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	"github.com/openshift/openstack-cinder-csi-driver-operator/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	storagelisters "k8s.io/client-go/listers/storage/v1"
+	"k8s.io/klog/v2"
+)
+
+// This CinderController watches OpenStack and:
+// 1) Installs Cinder CSI driver itself.
+// 2) Creates StorageClass for each volume type provided by Cinder.
+//
+// Note that StorageClasses are not deleted when a volumes type disappears
+// from Cinder.
+
+type CinderController struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Cinder is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Cinder and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 20 * time.Minute
+
+	operatorConditionPrefix = "CinderController"
+)
+
+func NewCinderController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &CinderController{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder.WithComponentSuffix("CinderController"),
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
```

Comment:
I'm not 100% on this API, but IIUC this is going to sync on any change to a storage class?

However (again IIUC), the controller doesn't appear to operate that way. We're really reconciling changes to OpenStack volume types, which we can't watch so we just sync every 20 minutes. Aren't storage classes immutable? I don't think there's any change which could be made to a storage class in K8s which we would be interested in. We're certainly not interested in the creation of storage classes.